crates/pathfinder/src/rpc/api.rs

```diff
@@ -331,6 +333,53 @@ impl RpcApi {
             .and_then(|x| x)
     }
 
+    /// Fetches [GlobalRoot] of a block by block hash.
+    ///
+    /// Tries storage first, then the sequencer.
+    async fn get_root_by_hash(&self, block_hash: StarknetBlockHash) -> RpcResult<GlobalRoot> {
+        let storage = self.storage.clone();
+
+        let handle = tokio::task::spawn_blocking(move || {
+            let mut connection = storage
+                .connection()
+                .context("Opening database connection")
+                .map_err(internal_server_error)?;
+
+            let transaction = connection
+                .transaction()
+                .context("Creating database transaction")
+                .map_err(internal_server_error)?;
+
+            let root = StarknetBlocksTable::get_root(
+                &transaction,
+                StarknetBlocksBlockId::Hash(block_hash),
+            )
+            .context("Read block from database")
+            .map_err(internal_server_error)?;
+
+            Ok(root)
+        });
+
+        let root = handle
+            .await
+            .context("Database read panic or shutting down")
+            .map_err(internal_server_error)
+            // flatten is unstable
+            .and_then(|x| x)?;
+
+        match root {
+            Some(root) => Ok(root),
+            None => Ok(self
+                .sequencer
+                .block_by_hash(BlockHashOrTag::Hash(block_hash))
+                .await
+                .context("Fetch block from sequencer")
+                .map_err(internal_server_error)?
+                .state_root
+                .expect("State root is always present for non pending block")),
```

Comment:
Since this is external data; I'm not sure we should panic if its missing?

Reply:
Sure but now we are assuming behaviour / coupling with a different component in our system. Which becomes dangerous.

In general, the fix is to have `pending` result be a different type so we can be more type safe. But as it is, I still wouldn't want to panic here.